### PR TITLE
[Issue 626] Add basic feature flag support to the API

### LINKED ISSUE #626: [Task]: Add basic feature flag for the API
### Summary

Feature flag functionality has been identified in the #70 . This task is to add the most basic feature flag functionality, to be determined by the implementer.

Feature flags has configurable defaults, and can be overridden upon different requests.

### Acceptance criteria

- [ ] Feature flag implemented
- [ ] Documentation added to `/documentation/api` 

### Questions / Notes
- Will it get updated in the template? Will make sure it scales well before updating the template. 
- Why do we need a feature flag? Want to change the behavior of an API, could use for front end as well; authn and authz will be added at some point and could use feature flags for that.

### PR DESCRIPTION
**Pull request by chouinar**
## Summary
Fixes #626

## Changes proposed
Added basic support for feature flags to the API with the following behavior:
* Ability to set a default value
* Ability to override the default with an environment variable (loaded at initialization of the application)
* Ability to pass in a header value and override the feature flag

Added a very basic feature flag at the moment that logs a message - will replace in the future once we have more realistic feature flags.

## Context for reviewers
This covers a lot of pretty reasonable use cases for feature flags in the API. The primary one that isn't yet supported is updating the feature flag on a running system, but the beginnings of that can be seen in this implementation (the initialize + global configuration approach could eventually allow for us to modify that global value periodically at run-time without adjusting the usage elsewhere).

## Additional information
You can see the header value working in the added test, but also can test it locally at localhost:8080/docs - a convenient dropdown is added for you.
Screenshot 2023-11-02 at 10 53 25 AM

### CODE REVIEW

**Comment by daphnegold** [comment]
What do you think about adding documentation on how developers add feature flags and reuse this code? A little blob in the README.md maybe?

File: api/openapi.generated.yml [hunk 1]
```diff
@@ -52,7 +52,13 @@ paths:
       summary: Health
   /v1/opportunities/search:
     post:
-      parameters: []
+      parameters:
+      - in: header
+        name: X-FF-Enable-Opportunity-Log-Msg
```
Comment (daphnegold) [inline api/openapi.generated.yml:57]: What a name :P

File: api/src/api/feature_flags/feature_flag.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    """
+
+    ### NOTE: This is a placeholder for future work, and only serves as a proof of concept of the idea.
+    #
+    # Header: X-FF-Enable-Opportunity-Log-Msg
+    # EnvVar: ENABLE_OPPORTUNITY_LOG_MSG
+    ENABLE_OPPORTUNITY_LOG_MSG = "enable_opportunity_log_msg"
+
+    def get_header_name(self) -> str:
+        value = "-".join([v.capitalize() for v in self.value.lower().split("_")])
```
Comment (daphnegold) [inline api/src/api/feature_flags/feature_flag.py:23]: Not clear, but aren't headers case-insensitive? What's the point of the funky capitalization + this logic to handle it?

**Review by daphnegold — APPROVED** [review]
:shipit: Great prototype!

File: documentation/api/feature-flags.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+Feature flags are setup in the API to allow for configuring behavior of the endpoints.
+
+The flags have a default value which can be adjusted by an environment variable per environment,
+and also allow for overriding it via headers in the API endpoints.
+
+## Naming Convention
+
+Naming of feature flags has the following convention:
+* Environment variables are always snake-case all-caps like `ENABLE_OPPORTUNITY_LOG_MSG`
+* Header fields are always prefixed with `X-FF` and are capitalized-kebab-cased like `X-FF-Enable-Opportunity-Log-Msg`
```
Comment (daphnegold) [inline documentation/api/feature-flags.md:12]: Why though?

File: api/src/api/opportunities/opportunity_schemas.py [hunk 4]
```diff
@@ -74,3 +79,26 @@ class OpportunitySearchSchema(request_schema.OrderedSchema):
         required=True,
     )
     paging = fields.Nested(PaginationSchema(), required=True)
+
+
+class OpportunitySearchHeaderSchema(request_schema.OrderedSchema):
```
Comment (acouch) [inline api/src/api/opportunities/opportunity_schemas.py:84]: This is outside the scope of this ticket, but when reviewing this, I'm noting we could [remove the OrderedSchema workaround](https://github.com/navapbc/template-application-flask/issues/59#issuecomment-1792425510) in a future ticket.

**Review by acouch — APPROVED** [review]
Awesome work!

File: documentation/api/feature-flags.md (hunk 3 above)
Comment (chouinar) [reply documentation/api/feature-flags.md:12]: Honestly, this is just following some behavior from other systems, so isn't something I thought too in-depth about. The capitalized-kebab-case approach seems to be the standard for headers. I didn't even know they were case-insensitive (and just tested that that is the case). Still prefer to follow the convention so our docs generate with the standard formatting.

Starting a header with `X-` seems to not be recommended much anymore (although it's not recommended against). Seems like the idea was that it was for internal values, but if they became a standard then it caused issues as people would not want the `X-` anymore. I don't think any of our variables will ever hit that case, but I can remove the `X-`. I like the `FF` bit to try to signal that it's overriding a feature in some way (although open to other naming conventions - this isn't something I'm super familiar with): https://datatracker.ietf.org/doc/html/rfc6648

File: api/src/api/feature_flags/feature_flag.py (hunk 2 above)
Comment (chouinar) [reply api/src/api/feature_flags/feature_flag.py:23]: Mainly so the docs generate with the normal convention

File: api/src/api/opportunities/opportunity_schemas.py (hunk 4 above)
Comment (chouinar) [reply api/src/api/opportunities/opportunity_schemas.py:84]: When we do the restructuring for error formatting of responses, I'll do it then. That approach will need a new base schema that we derive everything from to make some of the error formatting utils work happily.

**Pull request merged**